File: java/94.binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Deque<TreeNode> stack = new LinkedList<>();
   
        TreeNode help = root;
        while (help != null || !stack.isEmpty()) {
            while (help != null) {
                stack.offerFirst(help);
                help = help.left;
            }
            TreeNode node = stack.pollFirst();
            result.add(node.val);
            help = node.right;
        }
        return result;
    }
}

File: python3/94.binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        result = []
        helper = root
        stack = []
        while stack or helper is not None:
            while helper is not None:
                stack.append(helper)
                helper = helper.left
            curr = stack.pop()
            result.append(curr.val)
            if curr.right is not None:
                helper = curr.right
        return result